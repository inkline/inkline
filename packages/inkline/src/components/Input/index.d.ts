import { Component } from 'vue';

declare const IInput: Component;

export default IInput;
