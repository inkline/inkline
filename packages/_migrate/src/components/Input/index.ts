export { default as Input } from '@inkline/inkline/components/Input/Input.vue';
