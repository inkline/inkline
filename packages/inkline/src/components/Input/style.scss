@import '../../css/config/index';
@import '../../css/mixins/index';
@import '../IIcon/css/_variables';
@import '../IIcon/css/_mixins';
@import 'css/_variables';
@import 'css/_mixins';
@import 'css/_sizes';
@import 'css/_variants';

/**
 * Input
 */
.form-input {
    position: relative;
    display: flex;
    flex: 1 0 auto;
    flex-direction: row;

    @include form-input-selector() {
        position: relative;
        flex: 1 0 auto;
        width: 1%;
        padding: $input-padding-base;
        line-height: $input-line-height;
        background-clip: padding-box;
        border-width: $input-border-width;
        border-style: solid;
        box-shadow: $input-box-shadow;

        @if transition(input) {
            transition: $input-transition;
        }

        // Remove style for the caret on `<select>`s in IE10+.
        &::-ms-expand {
            background-color: transparent;
            border: 0;
        }

        // Customize the `:focus` state to imitate native WebKit styles.
        &:focus {
            outline: 0;

            // Avoid using mixin so we can pass custom focus shadow properly
            @if $box-shadow-enabled {
                box-shadow: $input-box-shadow;
            }
        }

        // Placeholder
        // Override Firefox's unusual default opacity see https://github.com/twbs/bootstrap/pull/11526.
        &::placeholder {
            opacity: 1;
        }

        // Disabled and read-only inputs
        //
        // HTML5 says that controls under a fieldset > legend:first-child won't be
        // disabled if the fieldset is disabled. Due to implementation difficulty, we
        // don't honor that edge case we style them as disabled anyway.
        &:disabled,
        &[readonly] {
            opacity: 1; // iOS fix for unreadable disabled content
            cursor: default;
        }

        /**
         * Select
         */
        &[readonly]:focus {
            outline: 0;
        }
    }

    .form-input-prefix,
    .form-input-suffix {
        z-index: 1;
        top: 50%;
        position: absolute;
        transform: translateY(-50%);
        font-style: normal;
        color: $text-muted;
        transition: border-color $transition-duration $transition-easing;

        > button {
            background: transparent;
            border: 0;
            color: $body-color;
        }
    }

    .form-input-prefix {
        left: $spacer;
        padding-right: $spacer;
        border-right-width: $input-prefix-border-width;
        border-right-style: solid;
    }

    .form-input-suffix {
        right: $spacer;
        padding-left: $spacer;
        border-left-width: $input-prefix-border-width;
        border-left-style: solid;
    }

    .clear-input {
        transition: background-color $transition-duration $transition-easing;
        display: flex;
        justify-content: center;
        align-items: center;
        cursor: pointer;
        width: $input-icon-circle-size;
        height: $input-icon-circle-size;
        border-radius: 100%;

        &::before {
            @extend %icon;

            font-size: 8px;
            content: $icon-times;
        }
    }

    @include form-input-size('md');
}

/**
 * Form Group
 */
.form-group.-error {
    .form-input-group > .form-input {
        @include form-input-selector() {
            border-color: colors('red');
        }
    }
}

/**
 * Form Input Group
 */
.form-input-group {
    display: block;
    vertical-align: middle;
    position: relative;

    &.-prepended,
    &.-appended {
        display: flex;
        flex-wrap: nowrap;

        .form-input {
            flex-shrink: 1;
        }
    }

    &.-prepended .form-input {
        @include form-input-selector() {
            border-top-left-radius: 0;
            border-bottom-left-radius: 0;
        }
    }

    &.-appended .form-input {
        @include form-input-selector() {
            border-top-right-radius: 0;
            border-bottom-right-radius: 0;
        }
    }

    > .form-input-prepend,
    > .form-input-append {
        display: flex;
        flex-flow: row nowrap;
        align-items: stretch;

        > * {
            flex: auto 1 0;
            border-radius: 0;
        }
    }

    @extend .form-input-group.-light;
    @extend .form-input-group.-md;
}

/**
 * Input Append and Prepend
 */
.form-input-prepend,
.form-input-append {
    align-items: center;
    display: flex;
    padding: 0;
    line-height: $input-line-height;
    transition: background-color $transition-duration $transition-easing;
}
