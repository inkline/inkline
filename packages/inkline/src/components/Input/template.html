<div class="form-input-group" :class="classes['root']">
    <div class="form-input-prepend" v-if="$slots.prepend">
        <slot name="prepend"></slot>
    </div>

    <div class="form-input"
         :class="classes['child']"
         @mouseenter="emitMouseEnter"
         @mouseleave="emitMouseLeave">

        <span class="form-input-prefix" v-if="$slots.prefix">
            <slot name="prefix"></slot>
        </span>

        <span class="form-input-suffix" v-if="$slots.suffix || clearable">
            <i class="clear-input" aria-label="Clear" v-show="isClearable" @click="clear"></i>
            <slot name="suffix"></slot>
        </span>

        <input
            v-bind="$attrs"
            v-model="model"
            ref="input"
            :name="name"
            :tabindex="tabIndex"
            :disabled="isDisabled"
            :aria-disabled="isDisabled"
            :readonly="isReadonly"
            :aria-readonly="isReadonly"
            @click="emitClick"
            @focus="emitFocus"
            @blur="emitBlur"
            @change="emitChange"
            @keydown="emitKeydown"
            @keyup="emitKeyup">
    </div>

    <div class="form-input-append" v-if="$slots.append">
        <slot name="append"></slot>
    </div>
</div>
