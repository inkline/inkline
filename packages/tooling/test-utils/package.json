{
    "name": "@inkline/test-utils",
    "version": "1.0.0",
    "type": "module",
    "scripts": {
        "dev": "devtools tsup watch",
        "build": "devtools tsup build",
        "lint": "devtools eslint"
    },
    "dependencies": {
        "@inkline/vue": "workspace:*"
    },
    "devDependencies": {
        "@inkline/devtools": "workspace:*",
        "@inkline/eslint-config": "workspace:*",
        "@inkline/types": "workspace:*",
        "@types/node": "catalog:",
        "@types/shelljs": "catalog:",
        "eslint": "catalog:",
        "tsx": "catalog:",
        "tsup": "catalog:",
        "typescript": "catalog:",
        "type-fest": "catalog:",
        "vue": "catalog:"
    },
    "peerDependencies": {
        "vue": ">=3.0.0"
    },
    "files": [
        "dist",
        "LICENSE"
    ],
    "types": "./dist/index.d.ts",
    "main": "./dist/index.js",
    "module": "./dist/index.cjs",
    "exports": {
        ".": {
            "types": "./dist/index.d.ts",
            "import": "./dist/index.js",
            "require": "./dist/index.cjs"
        },
        "./*": "./*"
    },
    "funding": {
        "type": "github",
        "url": "https://github.com/sponsors/alexgrozav"
    },
    "bugs": {
        "url": "https://github.com/inkline/inkline/issues"
    },
    "homepage": "https://inkline.io",
    "keywords": [
        "inkline",
        "ui",
        "ux",
        "dx",
        "library",
        "framework",
        "typescript"
    ],
    "engines": {
        "node": "*",
        "pnpm": "*"
    },
    "license": "ISC",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/inkline/inkline.git"
    }
}
