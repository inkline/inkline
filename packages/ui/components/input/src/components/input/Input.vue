<script lang="ts">
import type { PropType } from 'vue';
import { ref, computed, defineComponent, toRef, onBeforeUpdate } from 'vue';
import { filterKeys, uid } from '@inkline/utils';

import { Icon } from '@inkline/component-icon';
import type { FormStateKeys } from '@inkline/types';
import {
    useFormComponentColor,
    useFormComponentSize,
    useFormInputValidation
} from '@inkline/composables';

const componentName = 'Input';

export default defineComponent({
    name: componentName,
    components: {
        Icon
    },
    inheritAttrs: false,
    props: {
        /**
         * The color variant of the input
         * @type light | dark
         * @default
         * @name color
         */
        color: {
            type: String,
            default: undefined
        },
        /**
         * Display the input as clearable
         * @type Boolean
         * @default false
         * @name clearable
         */
        clearable: {
            type: Boolean,
            default: false
        },
        /**
         * The disabled state of the input
         * @type Boolean
         * @default false
         * @name disabled
         */
        disabled: {
            type: Boolean,
            default: false
        },
        /**
         * The error state of the input, computed based on schema by default.
         * @type Boolean | Array
         * @default ['touched', 'dirty', 'invalid']
         * @TODO use propDefaultValue to set default value
         * @name error
         */
        errorCondition: {
            type: [Boolean, Array] as PropType<boolean | FormStateKeys[]>,
            default: undefined
        },
        /**
         * The id of the internal input element
         * @type String
         * @default
         * @name id
         */
        id: {
            type: String,
            default: undefined
        },
        /**
         * The id of the input wrapper element
         * @type String
         * @default
         * @name wrapperId
         */
        wrapperId: {
            type: String,
            default: undefined
        },
        /**
         * Used to set the field value
         * @type String | Number
         * @default ''
         * @name modelValue
         */
        modelValue: {
            type: [String, Number],
            default: ''
        },
        /**
         * The unique identifier of the input
         * @type String
         * @default uid()
         * @name name
         */
        name: {
            type: String,
            default: () => {
                return uid('input');
            }
        },
        /**
         * Display the input as plaintext, disabling interaction
         * @type Boolean
         * @default false
         * @name plaintext
         */
        plaintext: {
            type: Boolean,
            default: false
        },
        /**
         * The readonly state of the input
         * @type Boolean
         * @default false
         * @name readonly
         */
        readonly: {
            type: Boolean,
            default: false
        },
        /**
         * The size variant of the input
         * @type sm | md | lg
         * @default
         * @name sizeMultiplier
         */
        size: {
            type: String,
            default: undefined
        },
        /**
         * The tabindex of the input
         * @type Number | String
         * @default 0
         * @name tabindex
         */
        tabindex: {
            type: [Number, String],
            default: 0
        },
        /**
         * The type of the input
         * @type String
         * @default text
         * @name type
         */
        type: {
            type: String,
            default: 'text'
        },
        /**
         * The aria-label of the clear button
         * @type String
         * @default Clear
         * @name clearAriaLabel
         */
        clearAriaLabel: {
            type: String,
            default: 'Clear'
        },
        /**
         * The aria-label of the show password toggle button
         * @type String
         * @default Toggle password visibility
         * @name showPasswordToggleAriaLabel
         */
        showPasswordToggleAriaLabel: {
            type: String,
            default: 'Toggle password visibility'
        },
        /**
         * Display the password toggle button
         * @type Boolean
         * @default true
         * @name showPasswordToggle
         */
        showPasswordToggle: {
            type: Boolean,
            default: true
        },
        /**
         * Enable input validation using schema
         * @type Boolean
         * @default true
         * @name validateSchema
         */
        shouldValidate: {
            type: Boolean,
            default: true
        }
    },
    emits: [
        /**
         * Event emitted for setting the modelValue
         * @type event
         * @name update:modelValue
         */
        'update:modelValue',
        /**
         * Event emitted when clearing the input element
         * @type event
         * @name clear
         */
        'clear',
        /**
         * Event emitted when toggling the password visibility
         * @type event
         * @name togglePassword
         */
        'togglePassword'
    ],
    setup(props, { attrs, emit, slots }) {
        const name = toRef(props, 'name');
        const disabled = toRef(props, 'disabled');
        const readonly = toRef(props, 'readonly');
        const shouldValidate = toRef(props, 'shouldValidate');
        const errorCondition = toRef(props, 'errorCondition');
        const {
            schema,
            hasError,
            isDisabled,
            isReadonly,
            onBlur: schemaOnBlur,
            onInput: schemaOnInput
        } = useFormInputValidation({
            name,
            disabled,
            readonly,
            errorCondition,
            shouldValidate
        });

        const rawColor = toRef(props, 'color');
        const rawSize = toRef(props, 'size');
        const { color } = useFormComponentColor(componentName, rawColor);
        const { size } = useFormComponentSize(componentName, rawSize);

        const input = ref<HTMLInputElement | null>(null);

        const wrapperAttrsAllowlist = ['class', 'className', /^data-/];
        const wrapperAttrs = computed(() =>
            filterKeys(attrs, { allowlist: wrapperAttrsAllowlist })
        );
        const inputAttrs = computed(() => filterKeys(attrs, { denylist: wrapperAttrsAllowlist }));

        const showPassword = ref(false);
        const isPasswordToggleable = computed(() => {
            return props.showPasswordToggle && props.type === 'password' && value.value !== '';
        });

        const inputType = computed(() => {
            if (props.type === 'password' && showPassword.value) {
                return 'text';
            }

            return props.type;
        });

        const tabIndex = computed(() => (isDisabled.value ? -1 : props.tabindex));

        const value = computed(() => {
            if (schema.value) {
                return schema.value.value;
            }

            return props.modelValue;
        });

        const isClearable = computed(() => {
            return props.clearable && !isDisabled.value && !isReadonly.value && value.value !== '';
        });

        const slotsClasses = ref(getSlotsClasses());
        const classes = computed(() => ({
            [`-${color.value}`]: true,
            [`-${size.value}`]: true,
            '-disabled': isDisabled.value,
            '-error': hasError.value,
            '-readonly': isReadonly.value,
            ...slotsClasses.value
        }));

        onBeforeUpdate(() => {
            slotsClasses.value = getSlotsClasses();
        });

        function getSlotsClasses() {
            return {
                '-prefixed': Boolean(slots.prefix),
                '-suffixed': Boolean(slots.suffix),
                '-prepended': Boolean(slots.prepend),
                '-appended': Boolean(slots.append)
            };
        }

        function onBlur(event: FocusEvent) {
            schemaOnBlur(event);
        }

        function onInput(event: Event) {
            const target = event.target as HTMLInputElement;

            schemaOnInput(target.value);
            emit('update:modelValue', target.value);
        }

        function onClear(event: Event) {
            event.stopPropagation();
            emit('clear', event);

            schemaOnInput('');
            emit('update:modelValue', '');
        }

        function onTogglePassword(event: Event) {
            event.stopPropagation();
            emit('togglePassword', event);

            showPassword.value = !showPassword.value;
        }

        return {
            input,
            wrapperAttrs,
            inputAttrs,
            tabIndex,
            value,
            inputType,
            isDisabled,
            isReadonly,
            isClearable,
            isPasswordToggleable,
            showPassword,
            classes,
            onBlur,
            onInput,
            onClear,
            onTogglePassword
        };
    }
});
</script>

<template>
    <div :id="id" class="input" :class="classes" v-bind="wrapperAttrs">
        <div v-if="$slots.prepend" class="input-prepend">
            <!-- @slot prepend Slot for the input prepend content -->
            <slot name="prepend" />
        </div>
        <div class="input-field">
            <span v-if="$slots.prefix" class="input-prefix">
                <!-- @slot prefix Slot for the input prefix content -->
                <slot name="prefix" />
            </span>
            <div v-if="$slots.value" class="input-value-overlay">
                <slot name="value" />
            </div>
            <input
                v-if="type !== 'textarea'"
                v-bind="inputAttrs"
                :id="id"
                ref="input"
                :value="value"
                :name="name"
                :type="inputType"
                :tabindex="tabIndex"
                :disabled="isDisabled"
                :aria-disabled="isDisabled ? 'true' : false"
                :readonly="isReadonly || plaintext"
                :aria-readonly="isReadonly || plaintext ? 'true' : false"
                @input="onInput"
                @blur="onBlur"
            />
            <textarea
                v-else
                v-bind="inputAttrs"
                :id="id"
                ref="input"
                :value="value"
                role="textbox"
                :name="name"
                :tabindex="tabIndex"
                :disabled="isDisabled"
                :aria-disabled="isDisabled ? 'true' : false"
                :readonly="isReadonly || plaintext"
                :aria-readonly="isReadonly || plaintext ? 'true' : false"
                aria-multiline="true"
                @input="onInput"
                @blur="onBlur"
            />
            <span v-if="$slots.suffix || isClearable || isPasswordToggleable" class="input-suffix">
                <!-- @slot password-toggle Slot for the password toggle button -->
                <slot
                    v-if="showPasswordToggle && type === 'password'"
                    name="password-toggle"
                    :on-toggle-password="onTogglePassword"
                >
                    <Icon
                        v-show="isPasswordToggleable"
                        :name="showPassword ? 'ink:eye' : 'ink:eye-off'"
                        class="input-icon input-password-toggle"
                        role="button"
                        :aria-label="showPasswordToggleAriaLabel"
                        :aria-hidden="isPasswordToggleable ? 'false' : 'true'"
                        @click="onTogglePassword"
                    />
                </slot>

                <!-- @slot clearable Slot for the clearable button -->
                <slot v-if="clearable" name="clearable" :on-clear="onClear">
                    <Icon
                        v-show="isClearable"
                        name="ink:close"
                        class="input-icon input-clear"
                        role="button"
                        :aria-label="clearAriaLabel"
                        :aria-hidden="isClearable ? 'false' : 'true'"
                        @click="onClear"
                    />
                </slot>

                <!-- @slot suffix Slot for the input suffix content -->
                <slot name="suffix" />
            </span>
        </div>
        <div v-if="$slots.append" class="input-append">
            <!-- @slot append Slot for the input append content -->
            <slot name="append" />
        </div>
    </div>
</template>
