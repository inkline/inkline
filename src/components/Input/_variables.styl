--label-margin-bottom := 0.5rem

--input-line-height := --line-height

--input-padding-sm := --spacers['1/2'] * --size-percentages['sm'] --spacers['1'] * --size-percentages['sm']
--input-padding-md := --spacers['1/2'] --spacers['1']
--input-padding-lg := --spacers['1/2'] * --size-percentages['lg'] --spacers['1'] * --size-percentages['lg']
--input-padding := {
    'sm': --input-padding-sm
    'md': --input-padding-md
    'lg': --input-padding-lg
}

--input-prefix-padding-sm := --input-padding-sm[1] + --input-padding-sm[1] * 3
--input-prefix-padding-md := --input-padding-md[1] + --input-padding-md[1] * 3
--input-prefix-padding-lg := --input-padding-lg[1] + --input-padding-lg[1] * 3
--input-prefix-padding := {
    'sm': --input-prefix-padding-sm
    'md': --input-prefix-padding-md
    'lg': --input-prefix-padding-lg
}

--input-font-size := --font-size

--input-color := --colors['gray-70']
--input-color-focus := --input-color
--input-placeholder-color := --colors['gray-60']
--input-plaintext-color := --body-color
--input-background := --colors['white']
--input-background-disabled := --colors['gray-20']
--input-background-focus := --input-background

--input-border-width := --border-width
--input-border-color := --border-color
--input-border-color-focus := darken(--input-border-color, 10%)
--input-border-radius := --border-radius

--input-box-shadow := inset 0 1px 1px rgba(--colors['black'], 0.05)
--input-transition := border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out

--input-icon-circle-size := 1.2rem
--input-icon-circle-background := --colors['gray-40']
--input-icon-circle-background-hover := --colors['gray-50']

--input-append-prepend-background-color := --colors['gray-30']
--input-append-prepend-background-gradient := --gradients['gray-30']
