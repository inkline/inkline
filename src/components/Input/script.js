import IFormGroup from '@inkline/inkline/src/components/FormGroup';

import AttributesProviderMixin from '@inkline/inkline/src/mixins/components/providers/AttributesProviderMixin';
import ClassesProviderMixin from '@inkline/inkline/src/mixins/components/providers/ClassesProviderMixin';
import InjectParentFormProviderMixin from '@inkline/inkline/src/mixins/forms/providers/InjectParentFormProviderMixin';
import ModelProviderMixin from '@inkline/inkline/src/mixins/forms/providers/ModelProviderMixin';
import SchemaProviderMixin from '@inkline/inkline/src/mixins/forms/providers/SchemaProviderMixin';

import ClickInputRefMethodMixin from '@inkline/inkline/src/mixins/forms/methods/ClickInputRefMethodMixin';
import FocusInputRefMethodMixin from '@inkline/inkline/src/mixins/forms/methods/FocusInputRefMethodMixin';
import EmitChangeMethodMixin from '@inkline/inkline/src/mixins/components/methods/EmitChangeMethodMixin';
import EmitClickMethodMixin from '@inkline/inkline/src/mixins/components/methods/EmitClickMethodMixin';
import EmitFocusMethodMixin from '@inkline/inkline/src/mixins/components/methods/EmitFocusMethodMixin';
import EmitHoverMethodMixin from '@inkline/inkline/src/mixins/components/methods/EmitHoverMethodMixin';
import EmitInputMethodMixin from '@inkline/inkline/src/mixins/components/methods/EmitInputMethodMixin';
import EmitKeydownMethodMixin from '@inkline/inkline/src/mixins/components/methods/EmitKeydownMethodMixin';

import ClearablePropertyMixin from '@inkline/inkline/src/mixins/forms/properties/ClearablePropertyMixin';
import DisabledPropertyMixin from '@inkline/inkline/src/mixins/forms/properties/DisabledPropertyMixin';
import NamePropertyMixin from '@inkline/inkline/src/mixins/forms/properties/NamePropertyMixin';
import ParentFormGroupPropertyMixin from '@inkline/inkline/src/mixins/forms/properties/ParentFormGroupPropertyMixin';
import ReadonlyPropertyMixin from '@inkline/inkline/src/mixins/forms/properties/ReadonlyPropertyMixin';
import SizePropertyMixin from '@inkline/inkline/src/mixins/components/properties/SizePropertyMixin';
import TabIndexPropertyMixin from '@inkline/inkline/src/mixins/components/properties/TabIndexPropertyMixin';

export default {
    name: 'IInput',
    inheritAttrs: false,
    components: {
        IFormGroup
    },
    mixins: [
        AttributesProviderMixin,
        ClassesProviderMixin,
        InjectParentFormProviderMixin,
        ModelProviderMixin,
        SchemaProviderMixin,

        ClickInputRefMethodMixin,
        FocusInputRefMethodMixin,
        EmitChangeMethodMixin,
        EmitClickMethodMixin,
        EmitFocusMethodMixin,
        EmitHoverMethodMixin,
        EmitInputMethodMixin,
        EmitKeydownMethodMixin,

        ClearablePropertyMixin,
        DisabledPropertyMixin,
        NamePropertyMixin,
        ParentFormGroupPropertyMixin,
        ReadonlyPropertyMixin,
        SizePropertyMixin,
        TabIndexPropertyMixin
    ],
    created() {
        this.classesProvider.add('root', () => ({
            '-prepended': Boolean(this.$slots.prepend) || this.prepended,
            '-appended': Boolean(this.$slots.append) || this.appended
        }));

        this.classesProvider.add('child', () => ({
            '-prefixed': Boolean(this.$slots.prefix),
            '-suffixed': Boolean(this.$slots.suffix)
        }));
    },
    mounted() {
        if (this.schema) {
            this.$set(this.parentFormGroup, 'inputSchema', this.schema);
        }
    }
};
