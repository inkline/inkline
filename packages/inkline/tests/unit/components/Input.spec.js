import { shallowMount } from '@vue/test-utils';
import Input from '@inkline/inkline/src/components/Input';

describe('Components', () => {
    describe('Input', () => {
        let wrapper;

        beforeEach(() => {
            wrapper = shallowMount(Input, {
                methods: {
                    created: Input.created,
                    mounted: Input.mounted
                },
                parentComponent: {
                    name: 'IFormGroup'
                }
            });
        });

        it('should be named correctly', () => {
            expect(Input.name).toEqual('IInput');
        });

        it('should render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('created()', () => {
            it('should add class rules to classes provider', () => {
                const classRulesLength = wrapper.vm.classesProvider.length;

                wrapper.vm.created();
                expect(wrapper.vm.classesProvider.length).toEqual(classRulesLength + 2)
            });

            it('should add "-prepended" class if "prepend" slot provided', () => {
                wrapper = shallowMount(Input, {
                    slots: {
                        prepend: ['<div />']
                    }
                });

                const rule = wrapper.vm.classesProvider[wrapper.vm.classesProvider.length - 2];

                expect(rule()).toEqual(expect.objectContaining({ '-prepended': true }));
            });

            it('should add "-appended" class if "append" slot provided', () => {
                wrapper = shallowMount(Input, {
                    slots: {
                        append: ['<div />']
                    }
                });

                const rule = wrapper.vm.classesProvider[wrapper.vm.classesProvider.length - 2];

                expect(rule()).toEqual(expect.objectContaining({ '-appended': true }));
            });

            it('should add "-prefixed" class if "prefix" slot provided', () => {
                wrapper = shallowMount(Input, {
                    slots: {
                        prefix: ['<div />']
                    }
                });

                const rule = wrapper.vm.classesProvider[wrapper.vm.classesProvider.length - 1];

                expect(rule()).toEqual(expect.objectContaining({ '-prefixed': true }));
            });

            it('should add "-suffixed" class if "suffix" slot provided', () => {
                wrapper = shallowMount(Input, {
                    slots: {
                        suffix: ['<div />']
                    }
                });

                const rule = wrapper.vm.classesProvider[wrapper.vm.classesProvider.length - 1];

                expect(rule()).toEqual(expect.objectContaining({ '-suffixed': true }));
            });
        });

        describe('mounted()', () => {
            it('should set input schema for parent form group', () => {
                const spy = jest.spyOn(wrapper.vm, '$set');

                wrapper.setProps({ schema: {} });
                wrapper.vm.mounted();

                expect(spy).toHaveBeenCalled()
            });
        });
    });
});
