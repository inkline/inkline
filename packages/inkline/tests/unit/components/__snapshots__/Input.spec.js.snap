// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Input should render correctly 1`] = `
<div class="form-input-group">
  <!---->
  <div class="form-input">
    <!---->
    <!----> <input tabindex="1"></div>
  <!---->
</div>
`;
